chore(TS): Convert Canvas to TS

asturur opened the pull request:
<!--
        Hi there!
        Thanks for taking the time and putting the effort into making fabric better! 💖
        Take a look at /CONTRIBUTING.md for crucial instructions regarding local setup, testing etc.
        https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md

        Adding tests that verify your fix and safeguard it from unwanted loss and changes is a MUST.

        Pull Requests are not always simple. Don't hesitate to ask for help (beware of [gotchas](http://fabricjs.com/fabric-gotchas) 😓).
        We appreciate your effort and would like the process to be productive and enjoyable.
        A strong community means a strong and better product for everyone.
-->
<!--
        📣 IMPORTANT NOTICE - PR LOCK DOWN 🔒    04/2022
        We are excited to announce that fabric is migrating to modern typescript/javascript 🤩.
        This means we will ⛔ not be accepting any PRs out of scope with the migration.
        We understand this might be annoying but wasted work is ever more so.
        The migration will be extreme on the source code so PRs from before will probably become stale to the point of death after the migration.
        It hurts us the throw away good work, effort and time put into fabric so please stay patient.
        You are welcome to join the migration effort 🔨
        https://github.com/fabricjs/fabric.js/issues/7596

        If you remain strong minded about PRing and the fix is small you can submit a PR to the 5.x branch
        During the migration we will port these changes to master
-->

## Motivation
Continuing conversion into TS.

<!-- Why you are proposing -->
<!-- You can use the @closes notation to mark issues that will be resolved by this PR -->

## Description

- just main class is changed, no mixin yet ( it took long )
- some chainability removed
-  _initInteractive now become _init, and also static_canvas got _init, in order to keep each class its own scope and avoid interactivity checks in static_canvas

## Changes

- Not initializing a Brush anymore when the canvas start ( seems a waste for doesn't brush at all and a concern for tree shaking )

<!-- before the fix vs. after -->

## Gist

<!-- Technical stuff if necessary -->

## In Action

<!-- Show case your accomplishment -->
<!-- Upload screenshots, screencasts and live examples showing your fix in contrast to the current state -->

ShaMan123 commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1567 @@
+   * @return {Boolean} true if the selection happened
+   * @private
+   */
+  _discardActiveObject(e?: MouseEvent, object?: FabricObject) {
```
TPointerEvent

asturur replied:
this was done

asturur commented:
@ShaMan123 i forked this branch and i m working on the canvas_events mixin, comment what you want to comment and i ll point to commits in that other branch that will address the comment

ShaMan123 commented:
I will review soon

ShaMan123 reviewed (commented):
Looks really good
My comment are negligent

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+  _init(el: string | HTMLCanvasElement, options = {}) {
+    this.renderAndResetBound = this.renderAndReset.bind(this);
+    this.requestRenderAllBound = this.requestRenderAll.bind(this);
+    this._initStatic(el, options);
```
because they do similar thing in different order and trying to change that approach broke some stuff.

ShaMan123 replied:
I suggest exposing `initElements` that will call `_createLowerCanvas` in StaticCanvas and be overridden by canvas.
then `_init` can be used with a super call.
Both the bindings and `calcOffset` seem to be redundant sine they are performed in StaticCanvas logic

ShaMan123 replied:
#8518

asturur replied:
That PR is fine i guess, that is what i tried to do, and maybe all my issue was the wrong import in the tests that took me  hours. When i fixed that i already rolled back the code and add no intention to fix it. Please open the PR on my newer branch https://github.com/fabricjs/fabric.js/pull/8519

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
-        this._initRetinaScaling();
+  /**
+   * During a mouse event we may need the pointer multiple times in multiple functions.
+   * _absolutePointer holds a reference to the pointer in coordinates that is valide for the event
```
i forgot to specify in which coordinates plane the pointer is.
`_absolutePointer holds a reference to the pointer in (canvas|fabricCanvas) coordinates that is valid for the event lifcycle`

ShaMan123 replied:
much better

asturur replied:
updated the descriptions in the other branch

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @param {boolean} altKey
+   * @returns {boolean} true if the transformation should be centered
+   */
+  _shouldCenterTransform(
```
yes the transform setup could be used to gather more information that are really useful, and less generic one. But not in this pr please

ShaMan123 replied:
for sure
I learned and now practice your approach

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+    let pointer = this.getPointer(e);
+    if (target.group) {
+      // transform pointer to target's containing coordinate plane
+      // should we use send point to plane?
```
ideally a point could have a tag or something that explains in what plane has been trasnformed the idea is useful, but is hard to put in practice because transform takes a generic matrix. Ma point.transform could at least take an extra parameters when is used with knowledge and that extra parameter could tag either the point type Point<Canvas> Point<Object> or add a string that represent the first 4 numbers of the transform matrix.
There are many things we could do to make geometry a little bit more intuitive even in documentation.

For example is not clear to everyone that image.width and height are in the same scale to cropX and cropY but left and top aren't at all.

ShaMan123 replied:
Good point

asturur replied:
moved to sendPointToPlane in the other branch

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @param {Boolean} skipGroup when true, activeGroup is skipped and only objects are traversed through
+   * @return {FabricObject | null} the target found
+   */
+  findTarget(e: TPointerEvent, skipGroup: boolean): FabricObject | null {
```
i think this has been addressed already in my new branch with default value to false

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+    // if interactive.
+    return target &&
+      isCollection(target) &&
+      target.interactive &&
```
this should be just subTargetCheck imho, we are returning the target, not selecting it directly.

ShaMan123 replied:
I agree
But where does the selection happen? In `findTarget`?

asturur replied:
wherever we do setActiveObject, Probably in the canvas_grouping mixin or half there and half somewhere else

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+    const wrapperEl = this.wrapperEl as HTMLDivElement,
+      lowerCanvasEl = this.lowerCanvasEl as HTMLCanvasElement,
+      upperCanvasEl = this.upperCanvasEl as HTMLCanvasElement,
+      cacheCanvasEl = this.cacheCanvasEl as HTMLCanvasElement;
```
sometimes it gets flagged in yellow, so i tend to prefer this one, but i can try again

ShaMan123 replied:
It is flagged because of tslint

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @type ModifierKey|ModifierKey[]
+   * @default
+   */
+  selectionKey: TOptionalModifierKey | ModifierKey[];
```
i know i don't remember who introduced it, but is recent and is only for selectionKey and not for the others.

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @type String
+   * @default
+   */
+  moveCursor: string;
```
i tried, but it turns out to be string anyway, it would be useful to do that anyway just i case TS then adds them

asturur replied:
This has been upgraded to `CSSStyleDeclaration['cursor'];` for a bunch of cursor i could find  here https://github.com/fabricjs/fabric.js/pull/8519

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
-        }
+  /**
+   * This is an internal method to decide if given the action and the modifier key pressed
+   * the transformation should with the object center as origin
```
yes bad english

asturur replied:
improved

```
   * This method will take in consideration a modifier key pressed and the control we are
   * about to drag, and try to guess the anchor point ( origin ) of the transormation.
   * This should be really in the realm of controls, and we should remove specific code for legacy
   * embedded actions.
```

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @private
+   * @deprecated
+   */
+  _getOriginFromCorner(
```
Yes i think this should be in the control realm too and out of the canvas class for sure

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+    this.wrapperEl.setAttribute('data-fabric', 'wrapper');
+    setStyle(this.wrapperEl, {
+      width: this.width + 'px',
+      height: this.height + 'px',
```
done

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+    // be transformed appropriately
+    // i.e. it should be serialised as it would appear if the selection group
+    // were to be destroyed.
+    const originalProperties = this._realizeGroupTransformOnObject(instance),
```
i would prefer understand how i can add the transform to SVG rather than removing the transfrom from the object.
Probably if i spend 20 minute on this i can improve it, i all boil down to add another matrix on the object.

asturur commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   * @param {FabricObject} [instance] the object to transform (gets mutated)
+   * @returns the original values of instance which were changed
+   */
+  _realizeGroupTransformOnObject(
```
yes after we merge this PR i ll do a mini PR to remove this particular thing here. Good call

asturur commented:
no it breaks the UTs as it was happening to me.

asturur commented:
@ShaMan123 i addressed all the comments that do not require a separate PR in the other branch.
If you don't have other comments i ll merge this tomorrow morning and continue from the canvas_events mixins

ShaMan123 commented:
> no it breaks the UTs as it was happening to me.

I found the commit that I pushed and didn't test that was the problem
#8520 is ready

ShaMan123 commented:
found a regression. no idea where my comments landed

ShaMan123 commented on src/canvas.class.ts:
```diff
@@ -1,1530 +1,1670 @@
+   */
+  _initRetinaScaling() {
+    super._initRetinaScaling();
+    this.__initRetinaScaling(this.upperCanvasEl, this.contextTop);
```
@asturur look here

ShaMan123 commented on src/static_canvas.class.ts:
```diff
@@ -378,22 +373,14 @@ export class StaticCanvas extends createCollectionMixin(
     if (!this._isRetinaScaling()) {
       return;
     }
```
the original
